### PR TITLE
DOCSP-37057 eloquent schema builder

### PR DESCRIPTION
**Pull request by ccho-mongodb**
JIRA: https://jira.mongodb.org/browse/DOCSP-37057
Staging:
- [Schema Builder](https://preview-mongodbcchomongodb.gatsbyjs.io/laravel/DOCSP-37057-eloquent-schema-builder/eloquent-models/schema-builder/)
- [Eloquent Models landing page](https://preview-mongodbcchomongodb.gatsbyjs.io/laravel/DOCSP-37057-eloquent-schema-builder/eloquent-models/)

### Checklist

- [ ] Add tests and ensure they pass
- [ ] Add an entry to the CHANGELOG.md file
- [ ] Update documentation for new features

### CODE REVIEW

File: docs/eloquent-models/schema-builder.txt [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+The following sections describe the Laravel schema builder features available
+in {+odm-short+} and show examples of how to use them:
+
+- :ref:`<laravel-eloquent-migrations>`
+- :ref:`<laravel-eloquent-collection-exists>`
+- :ref:`<laravel-eloquent-indexes>`
+
+.. note::
+
+   {+odm-short+} supports managing indexes and collection, but
```
Comment (jordan-smith721) [inline docs/eloquent-models/schema-builder.txt]:
```suggestion
   {+odm-short+} supports managing indexes and collections, but
```

File: docs/eloquent-models/schema-builder.txt [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+- Specify ``mongodb`` in the ``$connection`` field
+- Use only commands and syntax supported by {+odm-short+}
+
+.. tip::
+
+   Make sure you set ``DB_CONNECTION=mongodb`` in your ``.env`` configuration
+   file so that PHP artisan performs the migration on the correct database.
+
+The following example migration creates a collection and index when you run
+the migration and drops them when you roll back the migration:
```
Comment (jordan-smith721) [inline docs/eloquent-models/schema-builder.txt]: Suggestion: I had to read this a few times before I really understood what was happening. I think it could be easier to understand if we expanded on it a little bit (I'm not very well-versed in this topic so wording may need to be changed a little bit). Something like this maybe:
```suggestion
The following example creates a migration class with the following methods:

- ``up()``, which creates a collection and index when you run the migration.
- ``down()``, which drops the collection when you roll back the migration.
```

File: docs/eloquent-models/schema-builder.txt [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+   :end-before: end conditional create
+
+.. _laravel-eloquent-indexes:
+
+Manage Indexes
+--------------
+
+MongoDB indexes are data structures that improve query efficiency by reducing
+the number of documents needed to retrieve query results. Indexes such as
+geospatial indexes extend how you can query the data.
```
Comment (jordan-smith721) [inline docs/eloquent-models/schema-builder.txt]:
```suggestion
the number of documents needed to retrieve query results. Indexes, such as
geospatial indexes, extend how you can query the data.
```

File: docs/eloquent-models/schema-builder.txt [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+
+To improve query performance by using an index, make sure the index covers
+the query.  To learn more about indexes and query optimization, see the
+following MongoDB server manual entries:
```
Comment (jordan-smith721) [inline docs/eloquent-models/schema-builder.txt]: I think we decided on capitalizing "Server" for the Server docs (hope I'm remembering that right):
```suggestion
following MongoDB Server manual entries:
```

File: docs/eloquent-models/schema-builder.txt [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+          unique: true
+        }
+      ]
+
+Specify Index Options
+~~~~~~~~~~~~~~~~~~~~~
+
+MongoDB index options determine how the indexes are used and stored.
+You can specify index options when calling an index creation method such
+as ``index()`` on a ``Blueprint`` instance.
```
Comment (jordan-smith721) [inline docs/eloquent-models/schema-builder.txt]:
```suggestion
You can specify index options when calling an index creation method, such
as ``index()``, on a ``Blueprint`` instance.
```

File: docs/eloquent-models/schema-builder.txt [hunk 6]
```diff
@@ -0,0 +1,388 @@
...
+            maxVariable: 'punct',
+            normalization: false,
+            backwards: false,
+            version: '57.1'
+          }
+        }
+      ]
+
+To learn more about index options, see :manual:`Options for All Index Types </reference/method/db.collection.createIndex/#options-for-all-index-types>`
+in the MongoDB server manual.
```
Comment (jordan-smith721) [inline docs/eloquent-models/schema-builder.txt]:
```suggestion
in the MongoDB Server manual.
```

File: docs/eloquent-models/schema-builder.txt [hunk 7]
```diff
@@ -0,0 +1,388 @@
...
+
+Specify Index Options
+~~~~~~~~~~~~~~~~~~~~~
+
+MongoDB index options determine how the indexes are used and stored.
+You can specify index options when calling an index creation method such
+as ``index()`` on a ``Blueprint`` instance.
+
+The following migration code shows how to add a collation to an index as an
+index option. Click the :guilabel:`VIEW OUTPUT` button to see the indexes
```
Comment (jordan-smith721) [inline docs/eloquent-models/schema-builder.txt:206]: Suggestion: move this to its own line so it doesn't seem like it's supposed to be explaining what the example does further
```suggestion
index option. 

Click the :guilabel:`VIEW OUTPUT` button to see the indexes
```

File: docs/eloquent-models/schema-builder.txt [hunk 8]
```diff
@@ -0,0 +1,388 @@
...
+          expireAfterSeconds: 86400
+        }
+      ]
+
+You can specify sparse, TTL, and unique indexes on either a single field or
+compound index by specifying them in the index options.
+
+The following migration code shows how to create all three types of indexes
+on a single field. Click the :guilabel:`VIEW OUTPUT` button to see the indexes
+created by running the migration, including the default index on the ``_id``
```
Comment (jordan-smith721) [inline docs/eloquent-models/schema-builder.txt:297]: Same suggestion as above:
```suggestion
on a single field. 

Click the :guilabel:`VIEW OUTPUT` button to see the indexes
created by running the migration, including the default index on the ``_id``
```

File: docs/eloquent-models/schema-builder.txt [hunk 9]
```diff
@@ -0,0 +1,388 @@
...
+
+To create these index types, call the ``create()`` method on the ``Schema`` facade
+in your migration file. Pass ``create()`` the collection name and a callback
+method with a ``MongoDB\Laravel\Schema\Blueprint`` parameter. Call the
+appropriate helper method on the ``Blueprint`` instance and pass the
+index creation details.
+
+The following migration code shows how to create a sparse and TTL index
+by using the index helpers. Click the :guilabel:`VIEW OUTPUT` button to see
+the indexes created by running the migration, including the default index on
```
Comment (jordan-smith721) [inline docs/eloquent-models/schema-builder.txt:266]: Same suggestion as above:
```suggestion
by using the index helpers.

Click the :guilabel:`VIEW OUTPUT` button to see
the indexes created by running the migration, including the default index on
```

File: docs/eloquent-models/schema-builder.txt [hunk 10]
```diff
@@ -0,0 +1,388 @@
...
+- Unique indexes, which prevent inserting documents that contain duplicate
+  values for the indexed field
+
+To create these index types, call the ``create()`` method on the ``Schema`` facade
+in your migration file. Pass ``create()`` the collection name and a callback
+method with a ``MongoDB\Laravel\Schema\Blueprint`` parameter. Call the
+appropriate helper method on the ``Blueprint`` instance and pass the
+index creation details.
+
+The following migration code shows how to create a sparse and TTL index
```
Comment (jordan-smith721) [inline docs/eloquent-models/schema-builder.txt]:
```suggestion
The following migration code shows how to create a sparse and a TTL index
```

File: docs/eloquent-models/schema-builder.txt [hunk 11]
```diff
@@ -0,0 +1,388 @@
...
+inclusion, intersection, and proximity.
+
+To create geospatial indexes, call the ``create()`` method on the ``Schema`` facade
+in your migration file. Pass ``create()`` the collection name and a callback
+method with a ``MongoDB\Laravel\Schema\Blueprint`` parameter. Specify the
+geospatial index creation details on the ``Blueprint`` instance.
+
+The following example migration creates a ``2d`` and ``2dsphere`` geospatial
+index on the ``spaceports`` collection. Click the :guilabel:`VIEW OUTPUT`
+button to see the indexes created by running the migration, including the
```
Comment (jordan-smith721) [inline docs/eloquent-models/schema-builder.txt:340]:
```suggestion
index on the ``spaceports`` collection.

Click the :guilabel:`VIEW OUTPUT`
button to see the indexes created by running the migration, including the
```

File: docs/eloquent-models/schema-builder.txt [hunk 12]
```diff
@@ -0,0 +1,388 @@
...
+          key: { last_visible_dt: 1 },
+          name: 'last_visible_dt_1',
+          unique: true,
+          sparse: true,
+          expireAfterSeconds: 3600
+        }
+      ]
+
+To learn more about these indexes, see :manual:`Index Properties </core/indexes/index-properties/>`
+in the MongoDB server manual.
```
Comment (jordan-smith721) [inline docs/eloquent-models/schema-builder.txt]:
```suggestion
in the MongoDB Server manual.
```

File: docs/eloquent-models.txt [hunk 13]
```diff
@@ -6,517 +6,23 @@ Eloquent Models
...
 
 .. meta::
-   :keywords: php framework, odm, code example
+   :keywords: php framework, odm
 
-This package includes a MongoDB enabled Eloquent class that you can use to
-define models for corresponding collections.
+Eloquent models are part of the Laravel Eloquent object-relational
+mapping (ORM) framework that enable you to work with a database by using
+model classes. The {+odm-short+} extends this framework to use similar
```
Comment (jordan-smith721) [inline docs/eloquent-models.txt]:
```suggestion
model classes. {+odm-short+} extends this framework to use similar
```

File: docs/eloquent-models.txt [hunk 14]
```diff
@@ -6,517 +6,23 @@ Eloquent Models
...
+syntax to work with MongoDB as a database.
 
-Extending the base model
-~~~~~~~~~~~~~~~~~~~~~~~~
+This section contains guidance on how to use Eloquent models in
+{+odm-short+} to work with MongoDB in the following ways:
 
-To get started, create a new model class in your ``app\Models\`` directory.
+- :ref:`laravel-schema-builder` shows how to manage indexes on your MongoDB
+  collections
```
Comment (jordan-smith721) [inline docs/eloquent-models.txt]: Question: Should we mention that this page talks about migrations too?

```suggestion
- :ref:`laravel-schema-builder` shows how to perform migrations and manage indexes on your MongoDB
  collections
```

**Review by jordan-smith721 — CHANGES_REQUESTED** [review]
Nice work on a dense topic! Just a few suggestions and small fixes.

File: docs/eloquent-models/schema-builder.txt (hunk 4 above)
Comment (ccho-mongodb) [reply docs/eloquent-models/schema-builder.txt]: Thanks for bringing this up. I think it might be "Server manual" (omitting MongoDB, as in most of the driver docs) or "MongoDB server manual" (C# docs) from looking at the other docs repos. I'll use the former and create a source constant so that it's easy to change.

File: docs/eloquent-models/schema-builder.txt (hunk 7 above)
Comment (ccho-mongodb) [reply docs/eloquent-models/schema-builder.txt:206]: Thanks for the suggestion. I tried a few different revisions before deciding on this. I think we usually introduce the example in the paragraph directly before the example, so this seemed like the best option. I used this pattern in another related PR: https://github.com/mongodb/laravel-mongodb/pull/2747

Do you feel strongly about changing this? Let me know what you think.

File: docs/eloquent-models.txt (hunk 14 above)
Comment (ccho-mongodb) [reply docs/eloquent-models.txt]: I'll update this to "manage indexes on your collections by using Laravel migrations" since the content doesn't focus on migrations, but rather using them as a means to perform the index management (technically, you could use a different mechanism to manage the indexes).

File: docs/eloquent-models/schema-builder.txt (hunk 2 above)
Comment (ccho-mongodb) [reply docs/eloquent-models/schema-builder.txt]: Good idea. I had avoided this since the Laravel docs cover migrations in detail (linked on line 107), but seems like providing a list could help separate them.

File: docs/eloquent-models/schema-builder.txt (hunk 7 above)
Comment (ccho-mongodb) [reply docs/eloquent-models/schema-builder.txt:206]: Based on our Slack conversation, I'll leave it for now and make updates if any users complain.

**Review by jordan-smith721 — DISMISSED** [review]
LGTM!

File: docs/eloquent-models/schema-builder.txt [hunk 15]
```diff
@@ -0,0 +1,390 @@
...
+
+- :ref:`<laravel-eloquent-migrations>`
+- :ref:`<laravel-eloquent-collection-exists>`
+- :ref:`<laravel-eloquent-indexes>`
+
+.. note::
+
+   {+odm-short+} supports managing indexes and collections, but
+   excludes support for MongoDB JSON schemas for data validation. To learn
+   more about JSON schema validation, see :manual:`Schema Validation </core/schema-validation/>`
```
Comment (GromNaN) [inline docs/eloquent-models/schema-builder.txt:44]: Do you think we should add a method to update schema validation?
https://www.mongodb.com/docs/manual/core/schema-validation/update-schema-validation/#modify-the-validation-schema.

We have a feature request: [PHPORM-72](https://jira.mongodb.org/browse/PHPORM-72)

File: docs/eloquent-models/schema-builder.txt [hunk 16]
```diff
@@ -0,0 +1,390 @@
...
+Create a Migration Class
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can create migration classes manually or generate them by using the
+``php artisan make:migration`` command. If you generate them, you must make the
+following changes to perform the schema changes on your MongoDB database:
+
+- Replace the ``Illuminate\Database\Schema\Blueprint`` import with
+  ``MongoDB\Laravel\Schema\Blueprint`` if it is referenced in your migration
+- Specify ``mongodb`` in the ``$connection`` field
```
Comment (GromNaN) [inline docs/eloquent-models/schema-builder.txt]: The name of the connection don't have to be `mongodb`, it can be `default` or anything that is set in the configuration. You may document that "if your MongoDB connection is not the default one, then ...".

This is also for the `DB_CONNECTION` variable below.

File: docs/includes/schema-builder/astronauts_migration.php [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+
+declare(strict_types=1);
+
+use Illuminate\Database\Migrations\Migration;
+use Illuminate\Support\Facades\Schema;
+use MongoDB\Laravel\Schema\Blueprint;
+
+return new class extends Migration
+{
+    protected $connect = 'mongodb';
```
Comment (GromNaN) [inline docs/includes/schema-builder/astronauts_migration.php]: This property name is incorrect. Same for other examples.
```suggestion
    protected $connection = 'mongodb';
```

https://github.com/laravel/framework/blob/11.x/src/Illuminate/Database/Migrations/Migration.php#L12

File: docs/includes/schema-builder/flights_migration.php [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+{
+    protected $connect = 'mongodb';
+
+    public function up(): void
+    {
+      // begin create index
+        Schema::create('flights', function (Blueprint $collection) {
+            $collection->index('mission_type');
+            $collection->index(['launch_location' => 1, 'launch_date' => -1]);
+            $collection->unique('mission_id', null, null, ['name' => 'unique_mission_id_idx']);
```
Comment (GromNaN) [inline docs/includes/schema-builder/flights_migration.php]: Named arguments are useful here:
```suggestion
            $collection->unique('mission_id', options: ['name' => 'unique_mission_id_idx']);
```

File: docs/includes/schema-builder/passengers_migration.php [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+    protected $connect = 'mongodb';
+
+    public function up(): void
+    {
+        // begin index options
+        Schema::create('passengers', function (Blueprint $collection) {
+            $collection->index('last_name', 'passengers_collation_idx', null, [
+                'collation' => [ 'locale' => 'de@collation=phonebook', 'numericOrdering' => true ],
+            ]);
+        });
```
Comment (GromNaN) [inline docs/includes/schema-builder/passengers_migration.php]: With named arguments:
```suggestion
        Schema::create('passengers', function (Blueprint $collection) {
            $collection->index(
                'last_name',
                name: 'passengers_collation_idx', 
                options: [
                    'collation' => [ 'locale' => 'de@collation=phonebook', 'numericOrdering' => true ],
                ],
            );
        });
```

File: docs/includes/schema-builder/planets_migration.php [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+        // begin index helpers
+        Schema::create('planets', function (Blueprint $collection) {
+            $collection->sparse('rings');
+            $collection->expire('last_visible_dt', 86400);
+        });
+        // end index helpers
+
+        // begin multi index helpers
+        Schema::create('planet_systems', function (Blueprint $collection) {
+            $collection->index('last_visible_dt', null, null, ['sparse' => true, 'expireAfterSeconds' => 3600, 'unique' => true]);
```
Comment (GromNaN) [inline docs/includes/schema-builder/planets_migration.php]: With named arguments:
```suggestion
            $collection->index('last_visible_dt', options: ['sparse' => true, 'expireAfterSeconds' => 3600, 'unique' => true]);
```

**Review by GromNaN — CHANGES_REQUESTED** [review]

File: docs/eloquent-models/schema-builder.txt (hunk 15 above)
Comment (ccho-mongodb) [reply docs/eloquent-models/schema-builder.txt:44]: Sure, but probably a better question for Rishabh. Would there also be a command/option to create a schema validation?

File: docs/eloquent-models/schema-builder.txt [hunk 21]
```diff
@@ -0,0 +1,393 @@
...
+following changes to perform the schema changes on your MongoDB database:
+
+- Replace the ``Illuminate\Database\Schema\Blueprint`` import with
+  ``MongoDB\Laravel\Schema\Blueprint`` if it is referenced in your migration
+- Use only commands and syntax supported by {+odm-short+}
+
+.. tip::
+
+   If your default database connection is set to anything other than your
+   MongoDB database, update the following items so that PHP artisan performs
```
Comment (jmikola) [inline docs/eloquent-models/schema-builder.txt]: I'm not sure where "PHP artisan" was taken from (if it references the CLI command, note that "php" is only used to invoke the interpreter), but the command is referred to as proper noun Artisan in the Laravel docs: https://laravel.com/docs/11.x/artisan

File: docs/eloquent-models/schema-builder.txt [hunk 22]
```diff
@@ -0,0 +1,393 @@
...
+
+Manage Indexes
+--------------
+
+MongoDB indexes are data structures that improve query efficiency by reducing
+the number of documents needed to retrieve query results. Certain indexes, such
+as geospatial indexes, extend how you can query the data.
+
+To improve query performance by using an index, make sure the index covers
+the query.  To learn more about indexes and query optimization, see the
```
Comment (jmikola) [inline docs/eloquent-models/schema-builder.txt]:
```suggestion
the query. To learn more about indexes and query optimization, see the
```

I assume style guide calls for a single space for sentences.

File: docs/includes/schema-builder/flights_migration.php [hunk 23]
```diff
@@ -0,0 +1,32 @@
...
+use Illuminate\Support\Facades\Schema;
+use MongoDB\Laravel\Schema\Blueprint;
+
+return new class extends Migration
+{
+    protected $connection = 'mongodb';
+
+    public function up(): void
+    {
+      // begin create index
```
Comment (jmikola) [inline docs/includes/schema-builder/flights_migration.php]: The "begin" and "end" comments are not aligned.

File: src/Query/Builder.php [hunk 24]
```diff
@@ -1158,11 +1158,6 @@ protected function compileWheres(): array
         return $compiled;
     }
 
-    /**
-     * @param  array $where
-     *
-     * @return array
-     */
```
Comment (jmikola) [inline src/Query/Builder.php]: This change and `User.php` below seem unrelated to docs. How did these changes make it into the PR?

File: docs/eloquent-models/schema-builder.txt (hunk 22 above)
Comment (ccho-mongodb) [reply docs/eloquent-models/schema-builder.txt]: Good catch! Will fix. 
Since Snooty ignores repeated whitespace (unless used for indentation of RST elements), there's no choice/rule.

File: src/Query/Builder.php (hunk 24 above)
Comment (ccho-mongodb) [reply src/Query/Builder.php]: I think this was rebased in by the phpcbf GH action, but not really sure as I've never explicitly changed or committed the file.

Same change from another PR I branched from 4.1:
https://github.com/mongodb/laravel-mongodb/pull/2764/files#diff-bdf045bbd1faa398417a788236b3b7da49c553bc4c6e7d6403568ecd775d2034L1161-L1165

Let me know how best to proceed here.

Comment (GromNaN) [reply src/Query/Builder.php]: Yes, this cs fixes are automatically committed by the GHA. The PR needs to be rebased as the fix has been applied to the [branch 4.1](https://github.com/mongodb/laravel-mongodb/commits/4.1/).

Comment (ccho-mongodb) [reply src/Query/Builder.php]: Strange, when I used the GH button "Sync Fork" on this branch, it merged 4.2 into the branch. I'll revert that merge and perform the 4.1 merge manually.

File: docs/eloquent-models/schema-builder.txt [hunk 25]
```diff
@@ -0,0 +1,393 @@
...
+following changes to perform the schema changes on your MongoDB database:
+
+- Replace the ``Illuminate\Database\Schema\Blueprint`` import with
+  ``MongoDB\Laravel\Schema\Blueprint`` if it is referenced in your migration
+- Use only commands and syntax supported by {+odm-short+}
+
+.. tip::
+
+   If your default database connection is set to anything other than your
+   MongoDB database, update the following items so that Artisan performs
```
Comment (GromNaN) [inline docs/eloquent-models/schema-builder.txt]: This reference to "Artisan" is strange to me. The application is a unit (framework, config, classes...), so I'd use the passive form:

> update the following items so that the migration is performed on the correct database

File: docs/eloquent-models/schema-builder.txt [hunk 26]
```diff
@@ -0,0 +1,393 @@
...
+
+.. code-block:: bash
+
+   php artisan migrate --path=<path to your migration class file>
+
+This Artisan command runs the ``up()`` function in the class file to create
+the collection and index in the database specified in the ``config/database.php``
+file.
+
+To roll back the migration, run the following Artisan command after
```
Comment (GromNaN) [inline docs/eloquent-models/schema-builder.txt]:
```suggestion
To roll back the migration, run the following command after
```

File: docs/eloquent-models/schema-builder.txt [hunk 27]
```diff
@@ -0,0 +1,393 @@
...
+The following example migration drops an index called ``unique_mission_id_idx``
+from the ``flights`` collection:
+
+.. literalinclude:: /includes/schema-builder/flights_migration.php
+   :language: php
+   :dedent:
+   :start-after: begin drop index
+   :end-before: end drop index
+
+
```
Comment (GromNaN) [inline docs/eloquent-models/schema-builder.txt:393]: It would be interesting to add a section on search index creation. Waiting for a helper to be added: PHPORM-165

File: docs/upgrade.txt [hunk 28]
```diff
@@ -11,109 +11,39 @@ Upgrade Library Version
...
+- Replace calls to ``Query\Builder::whereAll($column, $values)`` with
+  ``Query\Builder::where($column, 'all', $values)``
 
-.. _laravel-breaking-changes:
+- ``Query\Builder::delete()`` doesn't accept ``limit()`` other than ``1`` or ``null``.
 
-Breaking Changes
-----------------
+- ``whereDate``, ``whereDay``, ``whereMonth``, ``whereYear``, ``whereTime``
+  now use MongoDB operators on date fields
```
Comment (GromNaN) [inline docs/upgrade.txt]: The changes in this file seems unrelated to the PR.

File: docs/eloquent-models/schema-builder.txt (hunk 27 above)
Comment (ccho-mongodb) [reply docs/eloquent-models/schema-builder.txt:393]: I created https://jira.mongodb.org/browse/DOCSP-37888 to track this.

I think you might be able to use the "Docs Changes Needed" checkbox in JIRA to generate tickets related to features. This workflow could be more convenient than communicating through PR comments or Slack.

File: docs/eloquent-models/schema-builder.txt (hunk 25 above)
Comment (ccho-mongodb) [reply docs/eloquent-models/schema-builder.txt]: Thanks, I'll avoid mentioning Artisan when possible.

File: docs/upgrade.txt (hunk 28 above)
Comment (ccho-mongodb) [reply docs/upgrade.txt]: Thanks for catching this -- I'll copy in the file from the 4.1 branch since it's not relevant to my changes.

**Review by GromNaN — APPROVED** [review]
LGTM
